sdk/include/platform/sunburst/platform-gpio.hh

```diff
@@ -1,78 +1,186 @@
-	uint32_t raspberryPiHeader;
-	uint32_t raspberryPiMask;
-	uint32_t arduinoShieldHeader;
-	uint32_t arduinoShieldMask;
+	uint32_t outputEnable;
+
+	/**
+	 * Returns the bit corresponding to a given GPIO index. Will mask out
+	 * the bit if this is outside of the provided mask.
+	 */
+	constexpr static uint32_t gpio_bit(uint32_t index, uint32_t mask)
+	{
+		return (1 << index) & mask;
+	}
+
+	/**
+	 * Set the output bit for a given GPIO pin index to a specified value.
+	 * This will only have an effect if the corresponding bit is first set
+	 * to `0` (i.e. output) in the `output_enable` register, and if the pin
+	 * is a valid output pin.
+	 */
+	void set_output(uint32_t index, bool value) volatile
+	{
+		const uint32_t Bit = gpio_bit(index, OutputMask);
+		if (value)
+		{
+			output = output | Bit;
+		}
+		else
+		{
+			output = output & ~Bit;
+		}
+	}
+
+	/**
+	 * Set the output enable bit for a given GPIO pin index. If `enable` is
+	 * true, the GPIO pin is set to output. If `false`, it is instead set to
+	 * input mode.
+	 */
+	void set_output_enable(uint32_t index, bool enable) volatile
+	{
+		const uint32_t Bit = gpio_bit(index, OutputEnableMask);
+		if (enable)
+		{
+			outputEnable = outputEnable | Bit;
+		}
+		else
+		{
+			outputEnable = outputEnable & ~Bit;
+		}
+	}
 
 	/**
-	 * Is the current target Sonata 1.0.0 or later?
+	 * Read the input value for a given GPIO pin index. For this to be
+	 * meaningful, the corresponding pin must be configured to be an input
+	 * first (set output enable to `false` for the given index). If given an
+	 * invalid GPIO pin (outside the input mask), then this value will
+	 * always be false.
 	 */
-	static bool constexpr Sonata1OrLater =
-#if SONATA >= CHERIOT_VERSION_TRIPLE(1, 0, 0)
-	  true
-#else
-	  false
-#endif
-	  ;
+	bool read_input(uint32_t index) volatile
+	{
+		return (input & gpio_bit(index, InputMask)) > 0;
+	}
+
+	/**
+	 * Read the debounced input value for a given GPIO pin index. FOr this
+	 * to be meaningful, the corresponding pin must be configured to be an
+	 * input first (set output enable to `false` for the given index). If
+	 * given an invalid GPIO pin (outside the input mask), then this value
+	 * will always be false.
+	 */
+	bool read_debounced_input(uint32_t index) volatile
+	{
+		return (debouncedInput & gpio_bit(index, InputMask)) > 0;
+	}
+};
+
+/**
+ * A driver for Sonata's Board GPIO (instance 0).
+ *
+ * Documentation source:
+ * https://lowrisc.org/sonata-system/doc/ip/gpio.html
+ */
+struct SonataGpioBoard : SonataGpioBase<0x0000'00FF, 0x0001'FFFF, 0x0000'0000>
+{
+	/**
+	 * Represents the state of Sonata's joystick, where each possible input
+	 * corresponds to a given bit in the General GPIO's input register.
+	 *
+	 * Note that up to 3 of these bits may be asserted at any given time:
+	 * pressing down the joystick whilst pushing it in a diagonal direction
+	 * (i.e. 2 cardinal directions).
+	 *
+	 */
+	enum class SonataJoystick : uint16_t
```

Comment:
```suggestion
	enum class [[clang::flag_enum]] SonataJoystick : uint16_t
```
As this can have more than one value we should tell the compiler that. I would also prefer the name `SonataJoystickDirection` or similar. `SonataJoystick` sounds to me like the actual GPIO register, which this isn't.

Reply:
I would actually like `read_joystick` to return a `SonataJoystickValue` class that has a variable of this enum as a member and convenience methods like `is_pressed()`.